File: src/main/java/com/alvcohen/travelhelper/api/AuthController.java
package com.alvcohen.travelhelper.api;

import com.alvcohen.travelhelper.dto.SignUpRequest;
import com.alvcohen.travelhelper.model.AuthProvider;
import com.alvcohen.travelhelper.model.User;
import com.alvcohen.travelhelper.repository.UserRepository;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.servlet.support.ServletUriComponentsBuilder;

import javax.validation.Valid;
import java.net.URI;

@RestController
@RequestMapping("/auth")
public class AuthController {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;

    public AuthController(UserRepository userRepository, PasswordEncoder passwordEncoder) {
        this.userRepository = userRepository;
        this.passwordEncoder = passwordEncoder;
    }

    @PostMapping("/signup")
    public ResponseEntity<?> registerUser(@Valid @RequestBody SignUpRequest signUpRequest) {
        if (userRepository.existsByEmail(signUpRequest.getEmail())) {
            return ResponseEntity.status(HttpStatus.CONFLICT).body(String.format("Email: %s already exists", signUpRequest.getEmail()));
        }

        User user = new User();
        user.setName(signUpRequest.getName());
        user.setEmail(signUpRequest.getEmail());
        user.setPassword(signUpRequest.getPassword());
        user.setProvider(AuthProvider.LOCAL);
        user.setPassword(passwordEncoder.encode(user.getPassword()));

        User result = userRepository.save(user);

        URI location = ServletUriComponentsBuilder
            .fromCurrentContextPath().path("/user/me")
            .buildAndExpand(result.getId()).toUri();

        return ResponseEntity.created(location).build();
    }


}

File: src/main/kotlin/com/alvcohen/travelhelper/controllers/AuthController.kt
package com.alvcohen.travelhelper.controllers

import com.alvcohen.travelhelper.AuthProvider
import com.alvcohen.travelhelper.User
import com.alvcohen.travelhelper.UserRepository
import com.alvcohen.travelhelper.VisaUsage
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.servlet.support.ServletUriComponentsBuilder
import javax.validation.Valid
import javax.validation.constraints.Email
import javax.validation.constraints.NotBlank

data class SignUpRequest(@field:NotBlank val name: String,
                         @field:Email @field:NotBlank val email: String,
                         @field:NotBlank val password: String)

class AuthController(@Autowired val userRepository: UserRepository, @Autowired val passwordEncoder: PasswordEncoder) {

    @PostMapping("/signup")
    fun registerUser(@RequestBody signUpRequest: @Valid SignUpRequest): ResponseEntity<*>? {
        if (userRepository.existsByEmail(signUpRequest.email)) {
            return ResponseEntity.status(HttpStatus.CONFLICT).body(java.lang.String.format("Email: %s already exists", signUpRequest.email))
        }

        val user = User(null,
                signUpRequest.email,
                signUpRequest.name,
                false,
                mutableSetOf(),
                mutableSetOf(),
                passwordEncoder.encode(signUpRequest.password),
                AuthProvider.LOCAL,
                null)

        val result = userRepository.save(user)
        val location = ServletUriComponentsBuilder
                .fromCurrentContextPath().path("/user/me")
                .buildAndExpand(result.id).toUri()
        return ResponseEntity.created(location).build<Any>()
    }

}